src/passes/TypeSSA.cpp

```diff
@@ -147,14 +148,136 @@ std::vector<HeapType> ensureTypesAreInNewRecGroup(RecGroup recGroup,
+using ExactTypes = std::unordered_set<HeapType>;
+
+struct Analyzer
+  : public PostWalker<Analyzer, UnifiedExpressionVisitor<Analyzer>> {
+  // Find allocations we can potentially optimize.
   News news;
 
-  void visitStructNew(StructNew* curr) { news.push_back(curr); }
-  void visitArrayNew(ArrayNew* curr) { news.push_back(curr); }
-  void visitArrayNewData(ArrayNewData* curr) { news.push_back(curr); }
-  void visitArrayNewElem(ArrayNewElem* curr) { news.push_back(curr); }
-  void visitArrayNewFixed(ArrayNewFixed* curr) { news.push_back(curr); }
+  // Also find heap types for which the exactness of allocations is observed. We
+  // will not be able to optimize allocations of these types without an analysis
+  // proving that an allocation does not flow into any location where its
+  // exactness is observed.
+  ExactTypes exactTypes;
+
+  // Find allocations we can potentially optimize.
+  void visitStructNew(StructNew* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNew(ArrayNew* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewData(ArrayNewData* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewElem(ArrayNewElem* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+  void visitArrayNewFixed(ArrayNewFixed* curr) {
+    news.push_back(curr);
+    visitExpression(curr);
+  }
+
+  // Find casts to exact types. Allocations of these types will not be able to
+  // be optimized.
+  template<typename Cast> void visitCast(Cast* cast) {
+    if (auto type = cast->getCastType(); type.isExact()) {
+      exactTypes.insert(type.getHeapType());
+    }
+  }
+  void visitRefTest(RefTest* curr) { visitCast(curr); }
+  void visitRefCast(RefCast* curr) { visitCast(curr); }
+  void visitBrOn(BrOn* curr) {
+    if (curr->op == BrOnCast || curr->op == BrOnCastFail) {
+      visitCast(curr);
+    }
+  }
+
+  void visitExpression(Expression* curr) {
+    // Look at the constraints on this expression's operands to see if it
+    // requires an exact operand. If it does, we cannot optimize allocations of
+    // that type. As an optimization, do not let control flow structures or
+    // branches inhibit optimization since they can safely be refinalized to use
+    // new types as long as no other instruction expected the original exact
+    // type. Also allow optimizing if the instruction that would inhibit
+    // optimizing will not be written in the final output.
+    if (Properties::isControlFlowStructure(curr) ||
+        Properties::isBranch(curr) ||
+        Properties::hasUnwritableTypeImmediate(curr)) {
+      return;
+    }
+
+    // Also do not let unreachable instructions inhibit optimization, as long as
+    // they are unreachable because of an unreachable child. (Some other
+    // unreachable instructions, such as a return_call, can still require an
+    // exact operand and may inhibit optimization.)
```

Comment:
This is perhaps not worth doing, as DCE will avoid such situations entirely?

Reply:
This is also necessary to avoid assertion failures in ChildTyper. ChildTyper asserts that the child determining the required types of the other children is not unreachable or null.